### DISCUSSION TITLE
Fading Image

### DISCUSSION THREAD

**Discussion by SaKiEQ** [Bodmer/TFT_eSPI#2863] [General]
Howdy, 

has anybody ever done a image fade, from to another, or from image to a color ? 

I am running on a ESP32, with 240x320 color display, so there is some processing power,
but it is also doing a lot of other things. I might be a low on memory, so soaring all the in-between
frames seems too much of a punishment. 

I have a compass dial, as a grey scale image, that I want to fade in color depending on ambient light.
All other screen elements worked fine, I just haven't talked the image portion yet. 

The image starts off a greyscale, the end would be all the white parts would be gradually changed to red. 
Would some form of XOR function on pixel level be performant enough, or is there a 'cheaper' way ? 

Any pointers, hints or examples ? 

Thx

**Comment by Bodmer** [comment]
The [alphaBlend function](https://github.com/Bodmer/TFT_eSPI/blob/master/TFT_eSPI.h#L732-L738) will return a colour that is a blend between any two colours. The ratio of colours is given by alpha.

There is an example in the library:
https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Generic/alphaBlend_Test

Here is an example that fades out an image stored in LittleFS.

[TFT_LittleFS_Fade_BMP.zip](https://github.com/Bodmer/TFT_eSPI/files/12731936/TFT_LittleFS_Fade_BMP.zip)

**Reply by SaKiEQ** [reply]
Your a super star ! Thank very much, will try in the morning. 

I have been thinking all evening that it could be rather simple.
The base image is essentially grey scale (white background with black letters at day-time), the at night white is replaced by red. 
So in my thinking I just had to ditch the green and the blue, leaving the red behind.

Right now I have 2 images for each state, ditching the red one, while replacing it an array of pixels, will keep the memory footprint about the same. Only at the expense of updating / recomputing the array during the fade transition.

Never done this before, but fascinating.

**Comment by SaKiEQ** [comment]
It worked like a charm ! I am actually shocked how fast it processes on a
ESP32.

On Wed, Sep 27, 2023 at 12:24 AM SaKiEQ ***@***.***> wrote:

> Your a super star ! Thank very much, will try in the morning.
>
> I have been thinking all evening that it could be rather simple.
> The base image is essentially grey scale (white background with black
> letters at day-time), the at night white is replaced by red.
> So in my thinking I just had to ditch the green and the blue, leaving the
> red behind.
>
> Right now I have 2 images for each state, ditching the red one, while
> replacing it an array of pixels, will keep the memory footprint about the
> same. Only at the expense of updating / recomputing the array during the
> fade transition.
>
> Never done this before, but fascinating.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Bodmer/TFT_eSPI/discussions/2863#discussioncomment-7117116>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIO2YOHORGYN3UT64YZDTP3X4NBZLANCNFSM6AAAAAA5IB6GPQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Sarah Kimberly Euschen
Production Designer
Westbay Lagoon
Doha, Qatar

https://themagicalworldofsakie.wordpress.com

**Comment by SaKiEQ** [comment]
and here is the result... The display is running at 8bit ,since there is really no graphics going on that require more. 
Never mind the occasional flicker in the lower section of the display, that is just a bug I havent weeded out yet. 

The ESP32 drives the 240x320 touch display, handles a GPS Neo-7M, MPU6050, RTC, Magnetometer, runs an SQLite database, Wifi, Bluetooth, MQTT services in the background, while refreshing the display at around 200ms, so approx 5fps. 
The are a number of icons that being rendered in the background, that currently not visible, since there was no GPS fix and no motion. 

So overall impressive that it works so well. 

https://github.com/Bodmer/TFT_eSPI/assets/35499064/8c64e6a7-8ef8-43f3-892c-0b0be081793e

**Comment by dslocum** [comment]
Trying to run the alphaBlend_test sketch and getting compile errors below.  I'm using a RP2040 and ILI9341 that is working fine with other TFT_eSPI projects.  Any ideas please?
------------
`In file included from C:\Users\dsloc\Documents\Arduino\libraries\TFT_eSPI\examples\Generic\alphaBlend_Test\alphaBlend_Test.ino:28:
C:\Users\dsloc\Documents\Arduino\libraries\TFT_eSPI/TFT_eSPI.h:719:19: warning: inline function 'uint16_t TFT_eSPI::alphaBlend(uint8_t, uint16_t, uint16_t)' used but never defined
  719 |   inline uint16_t alphaBlend(uint8_t alpha, uint16_t fgc, uint16_t bgc);
      |                   ^~~~~~~~~~
`

**Reply by Bodmer** [reply]
Download and install the latest master library on Github. Let me know if it works.

Let me kinow which version of the RP2040 board package you are using.

**Comment by dslocum** [comment]
Sorry for the delay getting back to you.
 
I installed the latest TFT_eSPI (2.5.34) from the Arduino Library Manager, and now I can’t even get example sketches running.  I do note that it now complains about “TOUCH_CS pin not defined”, but it proceeds to compile – just white screen.
 
Do I need to change a setting or config?

**Reply by dslocum** [reply]
Got it working again.  As I'm a bit new to this Arduino environment, I hadn't realized that loading a new version of the library would overwrite my user config settings.  Glad I had a Cloud backup of the original!

And - the new library fixed the original compiler error issue.  

Thank you!

**Reply by Bodmer** [reply]
Yes, that is a Gotcha. Normally Arduino libraries do not need configuration within the library but TFT_eSPI does. There are some tips here that may be of interest:
https://github.com/Bodmer/TFT_eSPI#tips

**Comment by dslocum** [comment]
Thank you sir.  This is a hell of a nice library. Very complete and solid.
 
I’ll be sure to check out that link soon.  Greatly appreciate your efforts and assistance.

**Comment by dslocum** [comment]
Hi again.
 
I realize this is not your product and probably unfair asking you, but I’m trying to get LVGL to work.  I already have your library working nicely, but I need some of the controls that LVGL offers.  It seems it is “supposed” to integrate with your library, but after hours – no DAYS – of trying, I can’t get a good compile of even the included “LVGL_Arduino” demo.  Tried my best to follow the many convoluted configuration examples, but without success.
 
Any help of pointers would be most welcome.
 
Doug
